tsl/src/remote/connection.c

```diff
@@ -1259,11 +1263,71 @@ remote_connection_open_with_options(const char *node_name, List *connection_opti
 	return conn;
 }
 
+/*
+ * Based on PG's GetUserMapping, but this version does not fail when a user
+ * mapping is not found.
+ */
+static UserMapping *
+get_user_mapping(Oid userid, Oid serverid)
+{
+	Datum datum;
+	HeapTuple tp;
+	bool isnull;
+	UserMapping *um;
+
+	tp = SearchSysCache2(USERMAPPINGUSERSERVER,
+						 ObjectIdGetDatum(userid),
+						 ObjectIdGetDatum(serverid));
+
+	if (!HeapTupleIsValid(tp))
+	{
+		/* Not found for the specific user -- try PUBLIC */
+		tp = SearchSysCache2(USERMAPPINGUSERSERVER,
+							 ObjectIdGetDatum(InvalidOid),
+							 ObjectIdGetDatum(serverid));
+	}
+
+	if (!HeapTupleIsValid(tp))
+		return NULL;
+
+	um = (UserMapping *) palloc(sizeof(UserMapping));
+#if PG12_GE
+	um->umid = ((Form_pg_user_mapping) GETSTRUCT(tp))->oid;
+#else
+	um->umid = HeapTupleGetOid(tp);
+#endif
+	um->userid = userid;
+	um->serverid = serverid;
+
+	/* Extract the umoptions */
+	datum = SysCacheGetAttr(USERMAPPINGUSERSERVER, tp, Anum_pg_user_mapping_umoptions, &isnull);
+	if (isnull)
+		um->options = NIL;
+	else
+		um->options = untransformRelOptions(datum);
+
+	ReleaseSysCache(tp);
+
+	return um;
+}
+
+/*
+ * Add user info (username and optionally password) to the connection
+ * options).
+ */
 static List *
-add_username_to_server_options(ForeignServer *server, Oid user_id)
+add_userinfo_to_server_options(ForeignServer *server, Oid user_id)
 {
 	const char *user_name = GetUserNameFromId(user_id, false);
 	List *server_options = list_copy(server->options);
+	const UserMapping *um = get_user_mapping(user_id, server->serverid);
```

Comment:
I think it is better to use the original function with PG_TRY here rather than copying the original function and having to maintain it.

Reply:
It is not possible to capture errors and resuming as if no error.